packages/block-editor/src/components/link-control/test/index.js

```diff
@@ -145,6 +145,29 @@ describe( 'Searching for a link', () => {
 		expect( lastSearchResultItemHTML ).not.toEqual( expect.stringContaining( 'URL' ) );
 	} );
 
+	it( 'should trim search term', async () => {
+		const searchTerm = '   Hello    ';
+
+		act( () => {
+			render( <LinkControl />, container );
+		} );
+
+		// Search Input UI
+		const searchInput = container.querySelector( 'input[aria-label="URL"]' );
+
+		// Simulate searching for a term
+		act( () => {
+			Simulate.change( searchInput, { target: { value: searchTerm } } );
+		} );
+
+		// fetchFauxEntitySuggestions resolves on next "tick" of event loop
+		await eventLoopTick();
+
+		const searchResultElements = container.querySelectorAll( '.block-editor-link-control__search-item-title mark' );
+
+		expect( searchResultElements ).toHaveLength( 3 );
```

Comment:
Thanks for adding the unit test, @sainthkh 🙌 

In this test, if I understand correctly, we want to check if the search term got trimmed but we're actually checking if (for an untrimmed term) the search returns correct results. Do you think it would be better if we only tested the actual trimming here? - sth. like `expect( searchTerm ).toBe( 'Hello' )`. Otherwise, the test doesn't seem to match the description which can potentially lead to false positives.

Reply:
This test checks the result: highlight the term Hello correctly. 

I think your idea is better for unit test. Mine is a bit like e2e test.

Reply:
As `searchText` is a local variable, we need to find a way to mock LInkControlSearchItem.

But after trying to mock LinkControlSearchItem, but failed because it somewhat violates Rules of Hooks.

```js
it( 'should trim search term 2', async () => {
	jest.resetModules();
	const mock = jest.fn( () => <></> );
	jest.mock( '@wordpress/compose/src/hooks/use-instance-id' );
	jest.mock( '../search-item', () => ( {
		default: mock,
	} ) );
	require( '../search-item' );
	const { default: MockedLinkControl } = require( '../index' );
	const searchTerm = '   Hello    ';

	act( () => {
		render( <MockedLinkControl />, container );
	} );

	// Search Input UI
	const searchInput = container.querySelector( 'input[aria-label="URL"]' );

	// Simulate searching for a term
	act( () => {
		Simulate.change( searchInput, { target: { value: searchTerm } } );
	} );

	// fetchFauxEntitySuggestions resolves on next "tick" of event loop
	await eventLoopTick();

	expect( mock ).toHaveBeenCalledTimes( 3 );

	for ( let i = 0; i < 3; i++ ) {
		expect( mock[ i ][ 0 ] ).toMatchObject( {
			searchTerm: 'Hello',
		} );
	}
} );
```

My first attempt is checking the mark tags because when the text is not trimmed, it doesn't bold the test results. 

I thought of testing it with `enzyme`, but I know it is not recommended.

How about changing the test name to `'should trim the search term and show the correct result'`?

Reply:
Let me take a look at this tomorrow. I wrote a lot of the tests so I'm keen to find where they could be improved.

Reply:
Sorry I've been delayed here and I'm trying to find time to get to this.

Reply:
Hi @sainthkh. I'm so sorry for the delay here. I've now found time to look into this for you.

As you've indicated your tests aren't actually asserting against user behaviour. This is primarily because in the tests we are mocking out the `fetchSearchSuggestions` handler and therefore no matter what you put in the `<input>` you will always get the same results. 

https://github.com/WordPress/gutenberg/blob/9d67638709562eeb49f8fa47f5d73fe6d6b619d7/packages/block-editor/src/components/link-control/test/index.js#L42

As a result, I'd recommend amending your test as follows:

```js
it( 'should trim search term', async () => {
    const searchTerm = '   Hello    ';

    act( () => {
        render( <LinkControl />, container );
    } );

    // Search Input UI
    const searchInput = container.querySelector(
        'input[aria-label="URL"]'
    );

    // Simulate searching for a term
    act( () => {
        Simulate.change( searchInput, { target: { value: searchTerm } } );
    } );

    // fetchFauxEntitySuggestions resolves on next "tick" of event loop
    await eventLoopTick();

    const searchResultTextHighlightElements = Array.from(
        container.querySelectorAll(
            '[role="listbox"] button[role="option"] mark'
        )
    );

    const invalidResults = searchResultTextHighlightElements.find(
        ( mark ) => mark.innerHTML !== 'Hello'
    );

    // Grab the first argument that was passed to the fetchSuggestions
    // handler (which is mocked out).
    const mockFetchSuggestionsFirstArg =
        mockFetchSearchSuggestions.mock.calls[ 0 ][ 0 ];

    // Given we're mocking out the results we should always have 4 mark elements.
    expect( searchResultTextHighlightElements ).toHaveLength( 4 );

    // Make sure there are no `mark` elements which contain anything other
    // than the trimmed search term (ie: no whitespace).
    expect( invalidResults ).toBeFalsy();

    // Implementation detail test to ensure that the fetch handler is called
    // with the trimmed search value. We do this because we are mocking out
    // the fetch handler in our test so we need to assert it would be called
    // correctly in a real world scenario.
    expect( mockFetchSuggestionsFirstArg ).toEqual( 'Hello' );
} );
```

To explain:

* We grab the `mark` tags from the search results.
* We confidence check there are 4 of them - possibly not needed though due to mock.
* We check to see that all of the mark's contain the trimmed string _only_.
* We check that our mocked search handler is being called with the _trimmed_ search term.

Note: I'm usually against asserting against mock calls as it's testing an implementation detial. However in this case due to our mock always returning the same results no matter what you search for, we do need to assert it is being called correctly (ie: with trimmed term).

In terms of actual implementation, I'm not sure what you have currently will actually work. Indeed testing in browser reveals that is the case as XHR requests are fired with the untrimmed search term.

To avoid confusion, I would recommend adding the `.trim()` call in the underlying `URLInput` component in the following locations:

* `onChange` - https://github.com/WordPress/gutenberg/blob/9d67638709562eeb49f8fa47f5d73fe6d6b619d7/packages/block-editor/src/components/url-input/index.js#L213
* `onFocus` (you'll see this method appear once you rebase from upstream `master`) - https://github.com/WordPress/gutenberg/blob/9d67638709562eeb49f8fa47f5d73fe6d6b619d7/packages/block-editor/src/components/url-input/index.js#L225

You could look to "normalise" this by creating a method `normalizeInputValue()` which would trim the value in both locations, but that's optional.

I hope that helps? I think this is a worthy and worthwhile change and I'd be pleased to help you get this merged.

Much appreciated.